lucene/core/src/test/org/apache/lucene/search/TestConstantScoreScorer.java

```diff
@@ -0,0 +1,215 @@
+  }
+
+  public void testTwoPhaseMatching_ScoreMode_COMPLETE_NO_SCORES() throws Exception {
+    testTwoPhaseMatching(ScoreMode.COMPLETE_NO_SCORES);
+  }
+
+  private void testTwoPhaseMatching(ScoreMode scoreMode) throws Exception {
+    try (TestConstantScoreScorerIndex index = new TestConstantScoreScorerIndex()) {
+      int doc;
+      ConstantScoreScorer scorer = index.constantScoreScorer(PHRASE_QUERY, 1f, scoreMode);
+
+      // "foo bar" match
+      doc = scorer.iterator().nextDoc();
+      assertThat(doc, equalTo(2));
+      assertThat(scorer.score(), equalTo(1f));
+
+      // should not reset iterator
+      scorer.setMinCompetitiveScore(2f);
+      assertThat(scorer.docID(), equalTo(doc));
+      assertThat(scorer.iterator().docID(), equalTo(doc));
+      assertThat(scorer.score(), equalTo(1f));
+      
+      // "foo not bar" will match the approximation but not the two phase iterator
+
+      // "foo bar foo" match
+      doc = scorer.iterator().nextDoc();
+      assertThat(doc, equalTo(5));
+      assertThat(scorer.score(), equalTo(1f));
+
+      doc = scorer.iterator().nextDoc();
+      assertThat(doc, equalTo(NO_MORE_DOCS));
+    }
+  }
+
+  public void testTwoPhaseMatching_ScoreMode_TOP_SCORES() throws Exception {
+    try (TestConstantScoreScorerIndex index = new TestConstantScoreScorerIndex()) {
+      int doc;
+      ConstantScoreScorer scorer = index.constantScoreScorer(PHRASE_QUERY, 1f, ScoreMode.TOP_SCORES);
+
+      // "foo bar" match
+      doc = scorer.iterator().nextDoc();
+      assertThat(doc, equalTo(2));
+      assertThat(scorer.score(), equalTo(1f));
+
+      scorer.setMinCompetitiveScore(2f);
+      assertThat(scorer.docID(), equalTo(doc));
+      assertThat(scorer.iterator().docID(), equalTo(doc));
+      assertThat(scorer.score(), equalTo(1f));
+
+      doc = scorer.iterator().nextDoc();
+      assertThat(doc, equalTo(NO_MORE_DOCS));
+    }
+  }
+
+  static class TestConstantScoreScorerIndex implements AutoCloseable {
+    private final Directory directory;
+    private final RandomIndexWriter writer;
+    private final IndexReader reader;
+
+    TestConstantScoreScorerIndex() throws IOException {
+      directory = newDirectory();
+
+      writer = new RandomIndexWriter(random(), directory,
+          newIndexWriterConfig().setMergePolicy(newLogMergePolicy(random().nextBoolean())));
+
+      for (String VALUE : VALUES) {
+        Document doc = new Document();
+        doc.add(newTextField(FIELD, VALUE, Field.Store.YES));
+        writer.addDocument(doc);
+      }
+      writer.forceMerge(1);
+
+      reader = writer.getReader();
+      writer.close();
+    }
+
+    ConstantScoreScorer constantScoreScorer(Query query, float score, ScoreMode scoreMode) throws IOException {
+      IndexSearcher searcher = newSearcher(reader);
+      Weight weight = searcher.createWeight(new ConstantScoreQuery(query), scoreMode, 1);
+      LeafReaderContext context = searcher.getIndexReader().leaves().get(0);
```

Comment:
let's assert that leaves.size() is 1?

Reply:
Sure, fixed in c5a83179ebd4a2775e24bd6518ea445cc2cb700e, thanks :+1: